dask/async.py

```diff
@@ -458,16 +463,28 @@ def fire_task():
             data = dict((dep, state['cache'][dep])
                         for dep in get_dependencies(dsk, key))
             # Submit
-            apply_async(execute_task, args=[key, dsk[key], data, queue,
-                                            get_id, raise_on_exception])
+            apply_async(execute_task,
+                        args=(key, marshall((dsk[key], data)),
+                              marshall, unmarshall,
+                              get_id, raise_on_exception),
+                        callback=queue.put)
+
+        def fill_pool_with_ready_tasks():
+            while state['ready'] and len(state['running']) < num_workers:
+                fire_task()
 
         # Seed initial tasks into the thread pool
-        while state['ready'] and len(state['running']) < num_workers:
-            fire_task()
+        fill_pool_with_ready_tasks()
 
         # Main loop, wait on tasks to finish, insert new ones
         while state['waiting'] or state['ready'] or state['running']:
-            key, res, tb, worker_id = queue.get()
+            key, res_info = queue.get()
+            try:
+                res, tb, worker_id = unmarshall(res_info)
+            except Exception as e:
```

Comment:
I'm not sure I follow. My thought was:
- Errors that occur on the remote worker -> `remote_exception`
- Errors that occur in the main process (where the deserialization happens) -> just re-raise after cleaning up the scheduler, since these didn't occur remotely.

Can you expand on what you mean by:

> There is a class of exceptions that occur remotely that err only on deserialization.

Are you referring to exceptions that can't be unpickled? In that case, I still think that we should raise the unpickling error instead of wrapping it in a remote exception. The failure is still in the main process.